index.bs

```diff
@@ -292,6 +292,42 @@ partial interface NeuralNetworkContext {
     Computes a 2-D convolution given input and filter 4-D tensors.
 </div>
 
+### gemm ### {#api-neuralnetworkcontext-gemm}
+<script type=idl>
+partial interface NeuralNetworkContext {
+  static Operand gemm(NeuralNetworkContext nn, Operand a, Operand b, Operand c, 
+                      optional float alpha = 1.0, optional float beta = 1.0, 
+                      optional boolean aTranspose = false, optional boolean bTranspose = false);
+};
+</script>
+<div algorithm=gemm>
+    **Arguments:**
+        - *nn*: The neural network context.
+        - *a*: an {{Operand}}. The first input 2-D tensor.
+        - *b*: an {{Operand}}. The second input 2-D tensor.
+        - *c*: an {{Operand}}. The third input 2-D tensor.
+        - *alpha*: an optional {{float}} scalar multiplier for the first input, default to 1.0.
+        - *beta*: an optional {{float}} scalar multiplier for the third input, default to 1.0.
+        - *aTranspose*: an optional {{boolean}} indicating if the first input should be transposed prior to calculating the output, default to false.
+        - *bTranspose*: an optional {{boolean}} indicating if the second input should be transposed prior to calculating the output, default to false.
+
+    **Returns:** an {{Operand}}. The output 2-D tensor that contains the calculated product of all the inputs.
+
+    Calculate the <a href="https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms#Level_3">general matrix multiplication of the Basic Liner Algebra Subprograms</a>. The calculation follows the expression `alpha * A * B + beta * C`, where `A`, `B`, and `C` are matrices, and `A` and `B` may optionally be transposed prior to the calculation.
+
+    This static operation can be generically implemented in term of other interface operations as follow. However, the user agent is free to implement the operation differently to suit their needs.
+
+    <pre highlight="js">
+    if (aTranspose)
+      a = nn.transpose(a);
+
+    if (bTranspose)
+      b = nn.transpose(b);
+
+    return nn.add(nn.matmul(nn.mul(alpha, a), b), nn.mul(beta, c));
```

Comment:
The current `mul` operator doesn't accept a float argument, e.g `alpha` and `beta` here.

Reply:
Thanks for the feedback. You're right, the float value needs to first be converted to a constant tensor. Since we don't support scalar tensors (do we?), I assume we need to turn it into a single-value tensor. I've added a constant method that helps with this. In ONNX, there is the `constant` op that does this, and in TF.JS there is `tf.scalar'. Please take a look at my new commit.

Reply:
We do support scalar tensors. In current spec, the scalar tensor with value `alpha` can be created by:
```js
nn.constant({type: ‘float32'}, new Float32Array([alpha]))
```

Reply:
I just think `nn.constant(float value)` is a lot more succinct. Also having to create an array of one element seems like an unnecessary overhead.

Reply:
It makes sense.